pkg/driver/oxide/oxide.go

```diff
@@ -0,0 +1,423 @@
+		if *f.target == "" {
+			return requiredOptionError(f.flag)
+		}
+	}
+
+	d.Host = opts.String(flagHost)
+	d.Token = opts.String(flagToken)
+	d.Project = opts.String(flagProject)
+	d.UserData = opts.String(flagUserData)
+	d.ExtraUserData = opts.String(flagExtraUserData)
+
+	d.DiskSize = opts.Int(flagDiskSize)
+	d.DiskImageId = opts.String(flagDiskImageId)
+	d.Memory = opts.Int(flagMemory)
+	d.Ncpus = opts.Int(flagNCpus)
+
+	d.VpcName = opts.String(flagVpc)
+	d.SubnetName = opts.String(flagSubnet)
+
+	d.BaseDriver.SSHUser = opts.String(flagSSHUser)
+	d.SSHKeys = opts.StringSlice(flagSSHPubKeys)
+
+	// base driver flags
+	d.BaseDriver.SSHPort = 22
+	d.SetSwarmConfigFromFlags(opts)
+
+	return
+}
+
+// PreCreateCheck is called to enforce pre-creation steps
+func (d *Driver) PreCreateCheck() error {
+	if len(d.userDataBase64())*3/4 > 32767 {
+		return fmt.Errorf("user data is too large")
+	}
+	return nil
+}
+
+func (d *Driver) Create() (err error) {
+	var ctx = context.Background()
+	var networkInterfaces []oxide.InstanceNetworkInterface
+	var sshKey *oxide.SshKey
+
+	if err = d.SetupClient(); err != nil {
+		return
+	}
+
+	if sshKey, err = d.SshSetup(); err != nil {
+		return
+	}
+
+	sshKeys := []oxide.NameOrId{
+		oxide.NameOrId(sshKey.Id),
+	}
+
+	for _, key := range d.SSHKeys {
+		sshKeys = append(sshKeys, oxide.NameOrId(key))
+	}
+
+	if _, err = d.Client.InstanceCreate(ctx,
+		oxide.InstanceCreateParams{
+			Project: oxide.NameOrId(d.Project),
+			Body: &oxide.InstanceCreate{
+				Description: "built by rancher-machine driver",
+				BootDisk: &oxide.InstanceDiskAttachment{
+					Description: "docker-machine disk",
+					DiskSource: oxide.DiskSource{
+						ImageId: d.DiskImageId,
+						Type:    oxide.DiskSourceTypeImage,
+					},
+					Name: oxide.Name(d.diskName()),
+					Size: oxide.ByteCount(d.DiskSize),
+					Type: oxide.InstanceDiskAttachmentTypeCreate,
+				},
+				Disks:    []oxide.InstanceDiskAttachment{},
+				Hostname: oxide.Hostname(d.MachineName),
+				Memory:   oxide.ByteCount(d.Memory),
+				Name:     oxide.Name(d.MachineName),
+				Ncpus:    oxide.InstanceCpuCount(d.Ncpus),
+				NetworkInterfaces: oxide.InstanceNetworkInterfaceAttachment{
+					Type: oxide.InstanceNetworkInterfaceAttachmentTypeCreate,
+					Params: []oxide.InstanceNetworkInterfaceCreate{
+						{
+							Description: "created by foo in rancher-machine driver",
+							Name:        "huh",
+							SubnetName:  oxide.Name(d.SubnetName),
+							VpcName:     oxide.Name(d.VpcName),
+						},
+					},
+				},
+				UserData:      d.userDataBase64(),
+				SshPublicKeys: sshKeys,
+			},
+		},
+	); err != nil {
+		return
+	}
+
+	if networkInterfaces, err = d.Client.InstanceNetworkInterfaceListAllPages(ctx,
+		oxide.InstanceNetworkInterfaceListParams{
+			Project:  oxide.NameOrId(d.Project),
+			Instance: oxide.NameOrId(d.MachineName),
+		},
+	); err != nil {
+		return
+	}
+
+	for _, nic := range networkInterfaces {
+		// TODO: not a foreach
```

Comment:
Is this "TODO" still necessary?

Reply:
This is kind of an open question. The network interfaces come back in order. The IPAdress needs to be the one that the rancher server can hit. so if there are multiple NICs, we may be able to check the VPC it's attached or subnet or something like that.. I'm adding this comment to the code actually :)